### DISCUSSION TITLE
How to Improve Request Rate

### DISCUSSION THREAD

**Discussion by PrathikshaGowda22** [fastapi/fastapi#8771] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import json
from typing import List, Dict
from async_lru import alru_cache
from fastapi import FastAPI
import aioredis

REDIS_HOST: str = "localhost"
app: FastAPI = FastAPI()

@alru_cache
async def get_redis() -> aioredis.Redis:
    """
    :return:
    """
    return await aioredis.create_redis_pool((REDIS_HOST, 6379), encoding='utf8')

@app.get('/redisConnection')
async def async_function():
    redis_conn: aioredis.Redis = await get_redis()
    return None

@app.get('/doOneRedisCall')
async def do_one_redis_call():
    redis_conn: aioredis.Redis = await get_redis()
    data: str = await redis_conn.get('random_key')
    return json.loads(data)

@app.get('/doMultipleRedisCall')
async def do_one_redis_call():
    redis: aioredis.Redis = await get_redis()
    redis_key_list: List[str] = ['random_key_1', 'random_key_2', 'random_key_3', 'random_key_4']
    response: List[Dict] = []
    for key in redis_key_list:  # making 4 get calls to redis
        data: str = await redis.get(key)
        response.append(json.loads(data))
    return response
```

### Description

We tested fastAPI application with Guicorn 20 workers, 

 Running this application using following cmd,
`gunicorn -w 20 -k uvicorn.workers.UvicornWorker --bind 0.0.0.0:80 main:app`

for /redisConnection API the machine was able to take load upto 7000 request per second
for /doOneRedisCall API the machine was able to take load upto 4700 request per second
for /doMultipleRedisCall API the machine was able to take load upto 1700 request per second

The more redis calls are made, less the request rate could be handled by the machine, what parameters in Gunicorn could we change, such that it could handle more request rate.?
We tried increasing
worker-connections from 1000(default) to 10000
keep-alive - 2(default) to 10 sec
backlog - 2048(default) to 100000
there is no improvement in request rate. Is there anything that i'm missing here .?
Thanks in advance .!!

### Operating System

Other

### Operating System Details

CPU : 8cores
RAM : 64GB
CPU MHz : 2700.516

### FastAPI Version

0.65.2

### Python Version

3.7

### Additional Context

pip freeze:
```
aioredis==1.3.1
asgiref==3.4.1
async-lru==1.0.2
async-timeout==3.0.1
click==8.0.1
fastapi==0.67.0
gunicorn==20.1.0
h11==0.12.0
hiredis==2.0.0
importlib-metadata==4.6.1
pydantic==1.8.2
redis==3.5.3
starlette==0.14.2
typing-extensions==3.10.0.0
uvicorn==0.14.0
zipp==3.5.0
```

**Comment by adriangb** [comment]
The main optimization you can make is making concurrent gets to redis in `doMultipleRedisCall`. Look into [`asyncio.create_task`](https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task).

**Comment by STeveShary** [comment, marked as answer]
@PrathikshaGowda22, You will want to make your async calls and "await" them in parallel, not serially (in a for loop).  You want to modify that function to the following:

```
import asyncio

@app.get('/doMultipleRedisCall')
async def do_one_redis_call():
    redis: aioredis.Redis = await get_redis()
    redis_key_list: List[str] = ['random_key_1', 'random_key_2', 'random_key_3', 'random_key_4']
    response: List[Dict] = []
    gathered_responses: List[str] = asyncio.gather([redis.get(key) for key in redis_key_list])
    for gathered_response in gathered_responses:
        response.append(json.loads(gathered_response))
    return response
```

More info here: https://docs.python.org/3/library/asyncio-task.html#running-tasks-concurrently
